### PR TITLE
Update blazor templates with Microsoft.Identity.Web APIs

### PR DESCRIPTION
**Pull request by JunTaoLuo**
# Description

Merging template changes from https://github.com/AzureAD/microsoft-identity-web/tree/master/ProjectTemplates/templates.
We are updating our Blazor Server and Blazor WASM templates to use Microsoft.Identity.Web APIs. The work presented here is similar to https://github.com/dotnet/aspnetcore/pull/24167 which updated our Mvc, RazorPages and Web API templates.

# Customer Impact

Current Blazor Server and Blazor WASM templates use obsoleted v1 AAD APIs. This change replaces their use in blazor templates with APIs from Microsoft.Identity.Web.

# Regression?

No

# Risk (see taxonomy)

Low. These are changes to templates only. The generated templates are verified manually (for functionality) and via unit tests (to ensure generated templates restores and builds).

# Link the PR to the original issue and/or the PR to master.

Tracking issue: https://github.com/dotnet/aspnetcore/issues/22726

### CODE REVIEW

**Comment by JunTaoLuo** [comment]
I feel a little uneasy with these templates changes since it looks like they were based off of 3.1 blazor templates. There were some minor changes between 3.1 and 5.0 it seems so we'll need to test the functionality a little more thoroughly.

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/dotnetcli.host.json [hunk 1]
```diff
@@ -67,6 +67,22 @@
     "NoHttps": {
       "longName": "no-https",
       "shortName": ""
+    },
+    "RazorRuntimeCompilation": {
```
Comment (JunTaoLuo) [inline src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/dotnetcli.host.json:71]: This isn't used in this template why is it present in https://github.com/AzureAD/microsoft-identity-web/blob/33d7bacef9bcbecd29e955535bfbbdda20f73f41/ProjectTemplates/templates/BlazorServerWeb-CSharp/.template.config/dotnetcli.host.json#L71-L74?

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json [hunk 2]
```diff
@@ -127,6 +127,46 @@
...
+            "Pages/ShowProfile.razor"
+            ]
+        },
+        {
+          "condition": "(GenerateGraph)",
+          "rename": {
+            "Shared/NavMenu.CallsMicrosoftGraph.razor": "Shared/NavMenu.razor"
+          },
+          "exclude": [
+            "Shared/NavMenu.NoGraphOrApi.razor"
```
Comment (JunTaoLuo) [inline src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json:159]: I'm a little confused here. Shouldn't NavMenu.CallsWebApi.razor be excluded as well? Vice versa for GenerateApi as well.

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json [hunk 3]
```diff
@@ -127,6 +127,46 @@
...
+          "condition": "(!GenerateApi)",
+            "exclude": [
+              "Services/DownstreamWebApi.cs",
+              "Pages/CallWebApi.razor"
+            ]
+        },
+        {
+          "condition": "(!GenerateApiOrGraph)",
+          "rename": {
+            "Shared/NavMenu.NoGraphOrApi.razor": "Shared/NavMenu.razor"
```
Comment (JunTaoLuo) [inline src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json:141]: Why aren't NavMenu.CallsMicrosoftGraph.razor and NavMenu.CallsWebApi.razor excluded?

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Controllers/WeatherForecastController.cs [hunk 4]
```diff
@@ -1,13 +1,23 @@
...
+using Microsoft.Extensions.Configuration;
+using Microsoft.Identity.Web;
+using System.Net;
+using System.Net.Http;
+#endif
+#if (GenerateGraph)
+using Microsoft.Graph;
+#endif
 using Microsoft.AspNetCore.Mvc;
+using Microsoft.Identity.Web.Resource;
```
Comment (JunTaoLuo) [inline src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Controllers/WeatherForecastController.cs:18]: This looks like it should only be needed when IndividualB2c or OrgAuth is specified right?

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs [hunk 5]
```diff
@@ -26,8 +19,10 @@
 using Microsoft.Extensions.Hosting;
 using System.Linq;
 #if (IndividualLocalAuth)
-using ComponentsWebAssembly_CSharp.Server.Data;
-using ComponentsWebAssembly_CSharp.Server.Models;
+using Microsoft.AspNetCore.Authentication.JwtBearer;
+#endif
+#if (GenerateGraph)
```
Comment (JunTaoLuo) [inline src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs:24]: At https://github.com/AzureAD/microsoft-identity-web/blob/master/ProjectTemplates/templates/ComponentsWebAssembly-CSharp/Server/Startup.cs#L23-L25 this logic is nested in IndividualLocalAuth. That looked wrong to me, can you explain why it should be nested?

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs [hunk 6]
```diff
@@ -26,8 +19,10 @@
 using Microsoft.Extensions.Hosting;
 using System.Linq;
 #if (IndividualLocalAuth)
-using ComponentsWebAssembly_CSharp.Server.Data;
-using ComponentsWebAssembly_CSharp.Server.Models;
+using Microsoft.AspNetCore.Authentication.JwtBearer;
```
Comment (JunTaoLuo) [inline src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs:22]: Curious why the changes under IndividualLocalAuth was made in https://github.com/AzureAD/microsoft-identity-web/blob/master/ProjectTemplates/templates/ComponentsWebAssembly-CSharp/Server/Startup.cs#L22

**Comment by jmprieur** [comment]
@JunTaoLuo : in src\ProjectTemplates\Web.ProjectTemplates\content\WebApi-CSharp\Startup.cs, we need to replace 
- `AddMicrosoftWebAppCallsWebApi` `AddMicrosoftWebApiCallsWebApi`

**Comment by JunTaoLuo** [comment]
Also before we can merge this fix, we'll also need to update to 0.2.1-preview.

**Comment by jmprieur** [comment]
@JunTaoLuo; in src\ProjectTemplates\Web.ProjectTemplates\content\StarterWeb-CSharp\Views\Home\Index.cshtml, there is a `@* #if`, but I don't see the `#endif*@`

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/dotnetcli.host.json (hunk 1 above)
Comment (jmprieur) [reply src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/dotnetcli.host.json:71]: I'm not sure about `RazorRuntimeCompilation`, @JunTaoLuo 
I copied the templates from ASP.NET core last week ... maybe it was present before?

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json (hunk 3 above)
Comment (jmprieur) [reply src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json:141]: I tried, but there the template engine got confused. So, in the end, after a lot of research, I had to split what is common (the rename), and then what is specific to GenerateGraph and GenerateApi ....
It seems like a bug in the template engine, btw ..

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Pages/CallWebApi.razor [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    <h2>API Result</h2>
+    @apiResult
+}
+
+@code {
+    private string apiResult;
+
+    protected override async Task OnInitializedAsync()
+    {
+        apiResult = await downstreamAPI.CallWebApiAsync();
```
Comment (jmprieur) [inline src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Pages/CallWebApi.razor:26]: We are working on making this work when there is a session cookie. Expect change in this particular file.

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Pages/ShowProfile.razor [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            {
+                byte[] photoByte = ((System.IO.MemoryStream)photoStream).ToArray();
+                photo = Convert.ToBase64String(photoByte);
+                this.StateHasChanged();
+            }
+
+        }
+        catch (Exception)
+        {
+            photo = null;
```
Comment (jmprieur) [inline src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Pages/ShowProfile.razor:70]: Same thing we'll have a change today

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Controllers/WeatherForecastController.cs (hunk 4 above)
Comment (jmprieur) [reply src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Controllers/WeatherForecastController.cs:19]: Yes, you are right!

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs (hunk 6 above)
Comment (jmprieur) [reply src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs:22]: This is to prepare the .AddAuthentication() pattern that Barry wants;

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs [hunk 9]
```diff
@@ -65,16 +60,36 @@ public void ConfigureServices(IServiceCollection services)
...
 #endif
 #if (OrganizationalAuth)
-#pragma warning disable CS0618 // Type or member is obsolete
-            services.AddAuthentication(AzureADDefaults.BearerAuthenticationScheme)
-                .AddAzureADBearer(options => Configuration.Bind("AzureAd", options));
-#pragma warning restore CS0618 // Type or member is obsolete
+            // Adds Microsoft Identity platform (AAD v2.0) support to protect this Api
+            services.AddMicrosoftWebApiAuthentication(Configuration, "AzureAd")
+#if (GenerateApiOrGraph)
+                    .AddMicrosoftWebAppCallsWebApi(Configuration,
```
Comment (jmprieur) [inline src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs:66]:
```suggestion
                    .AddMicrosoftWebApiCallsWebApi(Configuration,
```

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs [hunk 10]
```diff
@@ -65,16 +60,36 @@ public void ConfigureServices(IServiceCollection services)
...
+                                       Configuration.GetValue<string>("CalledApi:CalledApiUrl"));
+#endif
 #elif (IndividualB2CAuth)
-#pragma warning disable CS0618 // Type or member is obsolete
-            services.AddAuthentication(AzureADB2CDefaults.BearerAuthenticationScheme)
-                .AddAzureADB2CBearer(options => Configuration.Bind("AzureAdB2C", options));
-#pragma warning restore CS0618 // Type or member is obsolete
+            services.AddMicrosoftWebApiAuthentication(Configuration, "AzureAdB2C")
+#if (GenerateApi)
+                    .AddMicrosoftWebAppCallsWebApi(Configuration,
```
Comment (jmprieur) [inline src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs:83]:
```suggestion
                    .AddMicrosoftWebApiCallsWebApi(Configuration,
```

**Comment by jmprieur** [comment]
> I feel a little uneasy with these templates changes since it looks like they were based off of 3.1 blazor templates. There were some minor changes between 3.1 and 5.0 it seems so we'll need to test the functionality a little more thoroughly.

When were the 5.0 one added? I took these last week?
I'm fine with reconciliating.

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Pages/ShowProfile.razor [hunk 11]
```diff
@@ -0,0 +1,75 @@
+﻿@page "/showprofile"
+
+@using Microsoft.Graph
+@inject Microsoft.Graph.GraphServiceClient _graphServiceClient
```
Comment (javiercn) [inline src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Pages/ShowProfile.razor:4]: This needs to be capitalized `GraphServiceClient`

Comment (javiercn) [reply src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Pages/ShowProfile.razor:4]:
```suggestion
@inject Microsoft.Graph.GraphServiceClient GraphServiceClient
```

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Pages/ShowProfile.razor [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    </table>
+}
+
+@code {
+    User user;
+    string photo;
+
+    protected override async Task OnInitializedAsync()
+    {
+        user = await _graphServiceClient.Me.Request().GetAsync();
```
Comment (javiercn) [inline src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Pages/ShowProfile.razor:50]:
```suggestion
        user = await GraphServiceClient.Me.Request().GetAsync();
```

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Services/MicrosoftGraphServiceExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+        /// Adds the Microsoft Graph client as a singleton.
+        /// </summary>
+        /// <param name="services">Service collection.</param>
+        /// <param name="initialScopes">Initial scopes.</param>
+        /// <param name="graphBaseUrl">Base URL for Microsoft graph. This can be
+        /// changed for instance for applications running in national clouds</param>
+        public static IServiceCollection AddMicrosoftGraph(this IServiceCollection services,
+                                             IEnumerable<string> initialScopes,
+                                             string graphBaseUrl = "https://graph.microsoft.com/v1.0")
+        {
```
Comment (javiercn) [inline src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Services/MicrosoftGraphServiceExtensions.cs:23]: nit: We don't indent code like this in our templates. We prefer avoiding indentation all together and if necessary we just indent one tab/4spaces to the right

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json (hunk 3 above)
Comment (JunTaoLuo) [reply src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json:141]: I'll see why it's being confused since it looks rather wrong if there are extraneous files.

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/dotnetcli.host.json (hunk 1 above)
Comment (JunTaoLuo) [reply src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/dotnetcli.host.json:71]: I think it was present in 3.1 templates. I'll remove it.

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs (hunk 5 above)
Comment (JunTaoLuo) [reply src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs:24]: I'm going to assume this was just an oversight and apply what I think is the right fix but would like to still double check if this was intentional @jmprieur @jennyf19

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/dotnetcli.host.json (hunk 1 above)
Comment (pranavkm) [reply src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/dotnetcli.host.json:71]: Yeah, this might be fine to remove here. We'd only enabled this feature in the MVC and Razor Pages templates

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json (hunk 3 above)
Comment (JunTaoLuo) [reply src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json]: It seems to work for me.

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs (hunk 6 above)
Comment (JunTaoLuo) [reply src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs]: Is that something we need now or is it part of api improvements we're doing later?

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Services/MicrosoftGraphServiceExtensions.cs (hunk 13 above)
Comment (JunTaoLuo) [reply src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Services/MicrosoftGraphServiceExtensions.cs:23]: I'll leave the nits for later, There are lots of style fixes we plan to do as part of API cleanup in the future.

**Comment by JunTaoLuo** [comment]
@jmprieur @jennyf19 the 5.0 templates are what's in this repo: https://github.com/dotnet/aspnetcore/tree/master/src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp. But I've already highlighted the differences I saw between 3.1 and 5.0 templates via comments and resovled the known differences I found. As long as we test the functionality of the template changes, I think we're okay.

> in src\ProjectTemplates\Web.ProjectTemplates\content\StarterWeb-CSharp\Views\Home\Index.cshtml, there is a @* #if, but I don't see the #endif*@

Yea I couldn't figure out what the correct syntax was. If I had `#endif*@` in the file at the end it gets written out to the final file and isn't excluded as "templating code". I'm still unsure what the correct format should be but for now, it seems missing that endif generates the correct output so I'm rolling with it. I'll try to track down someone from the templating engine team to figure out how to resolve it later.

Also can you respond to https://github.com/dotnet/aspnetcore/pull/24268#discussion_r459864901 and https://github.com/dotnet/aspnetcore/pull/24268#discussion_r459865098 when you have the chance?

**Comment by jmprieur** [comment]
@JunTaoLuo did you update your IdentityWebTemplates repository? (and which branch) so that I can test more?

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs (hunk 5 above)
Comment (jmprieur) [reply src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs:25]: Yes, I agree this was wrong; thanks @JunTaoLuo

File: src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json (hunk 3 above)
Comment (jmprieur) [reply src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/.template.config/template.json]: Oh ok. thanks for fixing, then!

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs (hunk 6 above)
Comment (jmprieur) [reply src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs]: Yes, it is.

**Comment by JunTaoLuo** [comment]
A few updates:

I was able to run the BlazorServer-SingleOrg-Api template and was able to sign in:

image

However I didn't see instructions for BlazorServer-SingleOrg-Graph or BlazorServer-IndividualB2C-Api at https://microsofteur-my.sharepoint.com/:w:/r/personal/jmprieur_microsoft_com/_layouts/15/guestaccess.aspx?e=4%3AeYZCMo&at=9&share=ETED-1idfGBOjNo_L8aatSwBNE1mZE9ju9sap6K5r_wzcQ so I didn't verify those scenarios.

As for the BlazorWasm templates. apart from one bug I found, I figured out that the templates generated are [conditional on whether it's called on the command line or VS](https://github.com/dotnet/aspnetcore/blob/master/src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/.template.config/template.json#L31-L32). I fiddled with the logic and generated what I assume are the templates we would want to test but I'm still running into the `browser-wasm` TFM in the client project. @pranavkm @javiercn how are these templates supposed to work? How do we verify our changes made to the Server project of the `blazorwasm` template?

**Comment by JunTaoLuo** [comment]
FYI I updated the templates at https://github.com/JunTaoLuo/IdentityWebTemplates.

**Comment by jmprieur** [comment]
Thanks @JunTaoLuo : so we did not take ComponentsWebAssembly-CSharp for Preview 8 ?

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs [hunk 14]
```diff
@@ -65,16 +61,36 @@ public void ConfigureServices(IServiceCollection services)
...
-#pragma warning restore CS0618 // Type or member is obsolete
+            // Adds Microsoft Identity platform (AAD v2.0) support to protect this Api
+            services.AddMicrosoftWebApiAuthentication(Configuration, "AzureAd")
+#if (GenerateApiOrGraph)
+                    .AddMicrosoftWebApiCallsWebApi(Configuration,
+                                                   "AzureAd")
+                    .AddInMemoryTokenCaches();
+
+#else
+                    ;
```
Comment (javiercn) [inline src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs:72]: You need to put ` services.AddMicrosoftWebApiAuthentication(Configuration, "AzureAd")` inside the if and use ` services.AddMicrosoftWebApiAuthentication(Configuration, "AzureAd");` on the else.

Otherwise you generate 
```
services.AddMicrosoftWebApiAuthentication(Configuration, "AzureAd")
;
```

**Review by javiercn — APPROVED** [review]
It's impossible for me to determine the correctness of all these templates/combinations.

I'm signing-off based on the idea that the changes look reasonable (without taking into account the feedback provided offline) and that you have done due diligence verifying the scenarios manually.

Feel free to merge once all tests are passing and you are confident that the different combinations work, and we'll address any bug if needed post preview8.

Thanks for all your hard work to get this into preview8! It's much appreciated!

**Comment by jmprieur** [comment]
@JunTaoLuo : I've update https://microsofteur-my.sharepoint.com/:w:/r/personal/jmprieur_microsoft_com/_layouts/15/guestaccess.aspx?e=4%3AeYZCMo&at=9&share=ETED-1idfGBOjNo_L8aatSwBNE1mZE9ju9sap6K5r_wzcQ  for BlazorServer-SingleOrg-Graph 

The ComponentsWebAssembly-CSharp template is missing a couple of changes

In `CallWebApi.razor`, 
- you need to add the 3 lines:

@using Microsoft.Identity.Web
@inject IDownstreamWebApi downstreamAPI
@inject MicrosoftIdentityConsentAndConditionalAccessHandler ConsentHandler

https://github.com/AzureAD/microsoft-identity-web/blob/2a2a110ebf68d52afdc4ee53cb2a8de14b87d674/ProjectTemplates/templates/BlazorServerWeb-CSharp/Pages/CallWebApi.razor#L4-L7


- In OnInitializedAsync()
  ```CSharp
        try
        {
            apiResult = await DownstreamAPI.CallWebApiAsync();
        }
        catch (Exception ex)
        {
            ConsentHandler.HandleException(ex);
        }
  ```

In Startup.cs: please add `.AddMicrosoftIdentityConsentHandler();` after `services.AddServerSideBlazor()`
- in `public void ConfigureServices(IServiceCollection services)`
```CSharp
            services.AddRazorPages();
            services.AddServerSideBlazor()
                .AddMicrosoftIdentityConsentHandler();
            services.AddSingleton<WeatherForecastService>();
  ```

https://github.com/AzureAD/microsoft-identity-web/blob/2a2a110ebf68d52afdc4ee53cb2a8de14b87d674/ProjectTemplates/templates/BlazorServerWeb-CSharp/Pages/CallWebApi.razor#L28-L35

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs (hunk 14 above)
Comment (JunTaoLuo) [reply src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Server/Startup.cs:72]: I don't understand what you are saying. There are two #if, which one are you referring to. Also, I think the idea is to have just 
```
services.AddMicrosoftWebApiAuthentication(Configuration, "AzureAd")
;
```
if OrganizationAuth is configured but no API or Graph is configured? This is what's being done in the other templates. Is that wrong? cc @jmprieur

**Comment by JunTaoLuo** [comment]
@jmprieur it turns out that to test the BlazorWasm templates, you'll need to install the preview 7 SDK: https://dotnet.microsoft.com/download/dotnet/5.0. I was able to run the server project with it installed.
